org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/builderState/AbstractBuilderState.java

```diff
@@ -52,22 +58,48 @@
 	
 	@Inject
 	private IStorage2UriMapper storage2UriMapper;
+	
+	@Inject
+	private IWorkspace workspace;
 
 	private volatile boolean isLoaded = false;
 
-	public synchronized void load() {
+	public void load() {
 		if (!isLoaded) {
-			resourceDescriptionData = new ResourceDescriptionsData(persister.load());
-			if(storage2UriMapper instanceof IStorage2UriMapperExtension)
-				((IStorage2UriMapperExtension) storage2UriMapper).initializeCache();
-			isLoaded = true;
+			IWorkspaceRunnable runnable = new IWorkspaceRunnable() {
+				@Override
+				public void run(IProgressMonitor monitor) throws CoreException {
+					if (!isLoaded) {
+						resourceDescriptionData = new ResourceDescriptionsData(persister.load());
+						if(storage2UriMapper instanceof IStorage2UriMapperExtension)
+							((IStorage2UriMapperExtension) storage2UriMapper).initializeCache();
+						isLoaded = true;
+					}
+				}
+			};
+			try {
+				if (isWorkspaceLocked()) {
+					runnable.run(null);
+				} else {
+					workspace.run(runnable, null, IWorkspace.AVOID_UPDATE, null);
+				}
+			} catch(CoreException e) {
+				log.error(e.getMessage(), e);
+			}
+		}
+	}
+	
+	@SuppressWarnings("restriction")
+	private boolean isWorkspaceLocked() {
```

Comment:
In Storage2UriMapperJavaImpl the helper method is called isWorkspaceLockedByCurrentThread. Shall they both be named equally?
This makes use of internal API. I did not see a way to achieve the same without these internals. Would it make sense to request an API method for future use, e.g. on IWorkspace?

Reply:
good catch. and there should be one method only